File: frontend/src/services/productService.js
import axios from 'axios'

const API_URL = `${import.meta.env.VITE_API_URL}/products`

const productService = {
  /** Retorna todos los productos con filtros opcionales */
  getAll: async ({ categoryId, code } = {}) => {
    const params = {}
    if (categoryId) params.category_id = categoryId
    if (code) params.code = code
    const response = await axios.get(`${API_URL}/`, { params })
    return response.data
  },

  /** Retorna un producto por su ID */
  getById: async (id) => {
    const response = await axios.get(`${API_URL}/${id}`)
    return response.data
  },

  /** Crea un nuevo producto */
  create: async (data) => {
    const response = await axios.post(`${API_URL}/`, data)
    return response.data
  },

  /** Actualiza un producto existente */
  update: async (id, data) => {
    const response = await axios.put(`${API_URL}/${id}`, data)
    return response.data
  },

  /** Elimina un producto por su ID */
  delete: async (id) => {
    await axios.delete(`${API_URL}/${id}`)
  },
}

export default productService

File: backend/app/services/product_service.py
from fastapi import HTTPException, status
from app.repositories.product_repository import ProductRepository
from app.repositories.category_repository import CategoryRepository
from app.schemas.product_schema import ProductCreate, ProductUpdate
from app.models.product import Product


class ProductService:
    """Capa de lógica de negocio para productos."""

    def __init__(self, product_repository: ProductRepository, category_repository: CategoryRepository):
        # Repositorios inyectados desde el exterior (principio de inversión de dependencias)
        self.product_repository = product_repository
        self.category_repository = category_repository

    def get_all(self) -> list[Product]:
        """Retorna todos los productos."""
        return self.product_repository.get_all()

    def get_by_id(self, product_id: int) -> Product:
        """Retorna un producto por ID. Lanza 404 si no existe."""
        product = self.product_repository.get_by_id(product_id)
        if not product:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Producto con id {product_id} no encontrado."
            )
        return product

    def create(self, data: ProductCreate) -> Product:
        """Crea un nuevo producto. Valida código único y existencia de categoría."""
        # Valida que el código no esté en uso
        if self.product_repository.get_by_code(data.code):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Ya existe un producto con el código '{data.code}'."
            )
        # Valida que la categoría exista
        if not self.category_repository.get_by_id(data.category_id):
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Categoría con id {data.category_id} no encontrada."
            )
        return self.product_repository.create(data)

    def update(self, product_id: int, data: ProductUpdate) -> Product:
        """Actualiza un producto existente. Valida código único y existencia de categoría."""
        product = self.get_by_id(product_id)

        # Valida código único solo si se está cambiando
        if data.code and data.code != product.code:
            if self.product_repository.get_by_code(data.code):
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Ya existe un producto con el código '{data.code}'."
                )

        # Valida que la categoría exista solo si se está cambiando
        if data.category_id and not self.category_repository.get_by_id(data.category_id):
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Categoría con id {data.category_id} no encontrada."
            )

        return self.product_repository.update(product, data)

    def delete(self, product_id: int) -> None:
        """Elimina un producto. Lanza 404 si no existe."""
        product = self.get_by_id(product_id)
        self.product_repository.delete(product)